Data Warehouse acceptance tests for catalog, hive

tevesz opened the pull request:
Cover the dw environment, db catalog and hive virtual warehouse creations and deletions as part of acceptance testing.

vcsomor commented on resources/dw/resource_dw_acc_test.go:
```diff
@@ -95,8 +102,26 @@ func AwsDataLakePreCheck(t *testing.T) {
+		t.Skipf("Terraform acceptance testing requires either %s or %s environment variables to be set", cdp.CdpProfileEnvVar, cdp.CdpAccessKeyIdEnvVar)
+	}
+
+	if os.Getenv(cdp.CdpAccessKeyIdEnvVar) != "" {
+		if _, ok := os.LookupEnv(cdp.CdpPrivateKeyEnvVar); !ok {
```
What if the `CdpPrivateKeyEnvVar` is set but the `CdpAccessKeyIdEnvVar`?

tevesz replied:
That is fine, we only need the profile to be set for sure, I could replace this logic with 
```
if ak, pk := os.Getenv(cdp.CdpAccessKeyIdEnvVar), os.Getenv(cdp.CdpPrivateKeyEnvVar); (ak != "" && pk == "") || (ak == "" && pk != "") {
```

tevesz commented on cdpacctest/acctest.go:
```diff
@@ -11,7 +11,11 @@
 import (
+	"encoding/base64"
 	"fmt"
+	environmentoperations "github.com/cloudera/terraform-provider-cdp/cdp-sdk-go/gen/environments/client/operations"
+	environmentsmodels "github.com/cloudera/terraform-provider-cdp/cdp-sdk-go/gen/environments/models"
+	"github.com/stretchr/testify/assert"
```
I'm adding one, it is in progress

tevesz replied:
sorry, they hbave alinter, they don't have an import reviser

tevesz commented on cdpacctest/acctest.go:
```diff
@@ -95,11 +104,25 @@ provider "cdp" {
-func TestAccAwsProviderConfig() string {
-	return `
-provider "aws" {
+type awsProvider struct {
+	Profile string
+	Region  string
```
yes, I've noticed this too yesterday, will fix it

tevesz replied:
done

tevesz commented on resources/dw/resource_dw_acc_test.go:
```diff
@@ -95,8 +102,26 @@ func AwsDataLakePreCheck(t *testing.T) {
+	accountParams := cdpacctest.NewAwsAccountCredentials(cdpacctest.RandomShortWithPrefix(cdpacctest.ResourcePrefix))
+	accountParams.WithAccountID(t)
+	accountParams.WithExternalID(t)
+	accountParams.WithPolicy(t)
```
done

tevesz commented on cdpacctest/acctest.go:
```diff
@@ -126,3 +149,97 @@ func GetCdpClientForAccTest() *cdp.Client {
+	payload := response.GetPayload()
+	assert.NotNil(t, payload)
+	return payload
+}
```
done

tevesz commented on cdpacctest/acctest.go:
```diff
@@ -126,3 +149,97 @@ func GetCdpClientForAccTest() *cdp.Client {
+	AccountID     string
+	ExternalID    string
+	DefaultPolicy string
+}
```
done

tevesz commented:
I think I've addressed all comments @vcsomor and @gregito , thanks for the review! There is one thing I want to make sure about, the Aws Profile (or other aws cred variable) will be needed for the DW acceptance runs, to deploy the roles and policies, is there any change that I should add for this? Like support secret access keys?

vcsomor commented on cdpacctest/acctest.go:
```diff
@@ -95,11 +105,25 @@ provider "cdp" {
 }
-`
+
+func NewAwsProvider(profile, region string) awsProvider {
```
there is a bit of a contradiction here too: `NewAwsProvider` is exported but the type.

vcsomor replied:
we can fix it later

tevesz replied:
OK lets discuss this tomorrow

vcsomor commented on cdpacctest/acctest.go:
```diff
@@ -126,3 +150,103 @@ func GetCdpClientForAccTest() *cdp.Client {
+	defaultPolicy string
+}
+
+func NewAwsAccountCredentials(name string) *awsAccountCredentials {
```
same here

tevesz replied:
I've exported the type because I had to rebase and push becaue of the linter changes anyways

vcsomor commented:
> There is one thing I want to make sure about, the Aws Profile (or other aws cred variable) will be needed for the DW acceptance runs, to deploy the roles and policies, is there any change that I should add for this? Like support secret access keys?

I don't know if I got this right. We can add further possibilities once they are required. 
In general, the underlying AWS SDK knows the best how to initialize the client. The best way is to assign roles to the APP/TEST/CONTAINER/VM trying to access the AWS API and the SDK just initializes with the `LoadDefaultConfig`.
The next least good solution is to go with profiles, the most insecure is the accesskey/secretkey pair.

tevesz commented:
@vcsomor  this is a question because the acc test is run by a jenkins job, and that most probably already has a couple of options to handle the secrets. Technically my question is what kind of secrets are required by the jenkins infra.
